Support calling assumed length dummy procedure

jeanPerier opened the pull request:
Lowering was not able to compile the following kind of program:

```
subroutine foo(bar)
 character*(*) :: bar
 print *, bar()
end subroutine

character(10) function bar()
  bar = "abcdefghijkl"
end function

program test
  character(10) :: bar
  external :: bar
  call foo(bar)
end
```

This resulted in error `missing length for character`.

To support this kind of program, pass the result character dummy procedure as
character dummy objects: with its length. This matches nvfortran, ifort,
nag and xlf behavior (but not gfortran that does not support the above
code).

This is done by using fir.boxchar. The rational for using it instead of
manually mangling the length argument is that in call like
`foo(c1, bar, c2)` with `c1` and `c2` F77 like characters, the result
length of `bar` must be placed between the one of `c1` and `c2` after
all other arguments. fir.boxchar guarantees exactly that.

Three places are impacted:
- The function interface lowering (mustPassDummyProcedureAsBoxChar is
the central place that can completly disable this feature On and Off).
- When placing a call to a dummy procedure, use the fir.boxchar length
as a last resort if present in the interface.
- When passing a procedure designator, forward the length for character
function based on mustPassDummyProcedureAsBoxChar.

psteinfeld commented on flang/lib/Lower/CallInterface.cpp:
```diff
@@ -606,6 +606,38 @@ class Fortran::lower::CallInterfaceImpl {
+    // character object gives the expected behavior.
+
+    // This ABI is compatible with ifort, nag, nvfortran, and xlf, but not
+    // gfortran. Gfortran does not passes the length and is therefore unable to
```
Nit -- "does not pass" rather than "does not passes".

psteinfeld commented on flang/lib/Lower/ConvertExpr.cpp:
```diff
@@ -557,23 +567,56 @@ class ScalarExprLowering {
+      Fortran::evaluate::DynamicType resultType = proc.GetType().value();
+      if (const auto &lengthExpr = resultType.GetCharLength()) {
+        // The length expression may refer to dummy argument symbols that are
+        // meaningless whithout any actual arguments. Leave the length as
```
"without" rather than "whithout".

psteinfeld reviewed (commented):
A few nits ...

I can confirm that, aside from the nits, all builds, tests, and looks good.  But you should probably get someone more familiar with the code to approve.

schweitzpgi commented on flang/test/Lower/dummy-procedure-character.f90:
```diff
@@ -0,0 +1,172 @@
+    character(7) function bar1()
+    end function
+  end interface
+! CHECK:  %[[VAL_0:.*]] = fir.address_of(@_QPbar1) : (!fir.ref<!fir.char<1,?>>, index) -> !fir.boxchar<1>
```
The type should have a `7` in the length field, it seems.

jeanPerier replied:
I agree. I added a commit to this to ensure constant length is propagated (since the issue was not introduced by this patch, there is a few small modification to existing test cases).

schweitzpgi commented on flang/test/Lower/dummy-procedure-character.f90:
```diff
@@ -0,0 +1,172 @@
+  end interface
+! CHECK:  %[[VAL_0:.*]] = fir.address_of(@_QPbar1) : (!fir.ref<!fir.char<1,?>>, index) -> !fir.boxchar<1>
+! CHECK:  %[[VAL_1:.*]] = arith.constant 7 : i64
+! CHECK:  %[[VAL_2:.*]] = fir.convert %[[VAL_0]] : ((!fir.ref<!fir.char<1,?>>, index) -> !fir.boxchar<1>) -> !fir.ref<!fir.char<1,?>>
```
This cast is lossy and odd looking to me. It is lossy in that we have a function pointer and are converting it to a data pointer, which loses information that the pointer is to code and not data. Subsequently, the IR has no way to reason about this value correctly.

jeanPerier replied:
Makes sense, what do you see as an alternative here ?  It is currently not possible to create a fir.embox_char with a function address. Should we allow that to avoid the cast or should we create a new operation here (other alternatives welcomed if you see any) ?

jeanPerier commented on flang/lib/Lower/ConvertExpr.cpp:
```diff
@@ -541,10 +541,20 @@ class ScalarExprLowering {
 
+  /// Does \p expr refer to symbols that are not mapped to IR values in the
+  /// currrent scope ?
+  bool onlyRefersKnownSymbols(const Fortran::lower::SomeExpr &expr) {
```
Sure, I moved it outside of the class and added a SymbolMap argument to make that possible. Also renamed as suggested.

jeanPerier reviewed (commented):
Thanks @schweitzpgi and @psteinfeld for the review. I have updated all the code comments and applied the small modifications.

@schweitzpgi I understand you are against using fir.boxchar as a tuple <function address , result length if know in the siganture> ? I can look into using an actual mlir tuple type here to keep the function type visible in the tuple. However this will require deeper change: A new case to handle in argument lowering and a modification in codegen to deal with the tuple like fir.boxchar<> (I liked the idea of using fir.boxchar to avoid having to add special handling in codegen). Does that sounds a better solution to you ?

jeanPerier commented:
@schweitzpgi , I have updated the PR to lower the character dummy procedure argument to a `tuple<function address, length value>` and added an optimizer patch that takes care of splitting this like with characters (It is in its own commit and has its own test so that we can upstream it directly).

Note that the function address type in the tuple remains poor (`() -> ()`) because so far we do not type dummy procedure as good as we could (this is not directly related to this PR: https://github.com/flang-compiler/f18-llvm-project/blob/94aa4163fd555b15f491d5728f64ef86cc2d32bb/flang/lib/Lower/CallInterface.cpp#L38).

jeanPerier commented:
Rebased to solve conflicts and added comments in Character.h.

schweitzpgi reviewed (approved):
Gave it a quick look. LGTM.

jeanPerier commented:
Patch for optimizer related change in LLVM: https://reviews.llvm.org/D118108
